### ISSUE TITLE
[competitor-port] Feature engineering: expression engine + cross-sectional processors + PIT data handler (qlib patterns)

### ISSUE THREAD

**Issue by KCNyu** [KCNyu/clawock#1142]
## 竞品做法
**qlib** — `qlib/data/ops.py` + `qlib/data/dataset/handler.py` + `qlib/data/dataset/processor.py` + `qlib/data/pit.py`（源码：https://github.com/microsoft/qlib/tree/main/qlib/data）

- **表达式引擎**：`ExpressionOps` 基类 + `OpsWrapper` 注册机制。核心算子分类：
  - **元素级**：`Abs`/`Sign`/`Log`/`Mask`/`Not`/`NpElemOperator` (np.abs/sign/log/mask/not)
  - **配对级**：`Add`/`Sub`/`Mul`/`Div`/`Greater`/`Less`/`Gt`/`Ge`/`Lt`/`Le`/`Eq`/`Ne`/`And`/`Or`/`Power` (np.add/subtract/multiply/divide/maximum/minimum/greater/greater_equal/less/less_equal/equal/not_equal/logical_and/logical_or/power)
  - **滚动/展开**：`Rolling` 基类 + `Sum`/`Mean`/`Std`/`Var`/`Skew`/`Kurt`/`Med`/`Mad`/`Slope`/`Rsquare`/`Resi`/`Rank`/`Quantile`/`Count`/`EMA`/`WMA`/`Delta`/`Min`/`Max`/`Corr`/`Cov`/`PairRolling` 子类
  - **时序重采样**：`TResample` (pandas resample)
  - **条件/特殊**：`If`/`Ref` (前向引用) /`Feature`/`PFeature`/`P` (点在时间坍缩) /`PRef` (指定周期回看)
  - **注册机制**：`OpsWrapper.register()` 支持自定义算子，`Operators` 全局单例

- **数据处理器**：`DataHandlerLP` 三层数据（DK_R 原始/DK_I 推理/DK_L 学习），分离学习/推理预处理流水线
  - `fetch()` 统一接口：`selector`/`level`/`col_set`/`data_key`/`proc_func`（预处理钩子）
  - `_fetch_data()` 支持 `proc_func` 预处理、多级索引切片、`squeeze` 压缩
  - `get_range_iterator()` 滑动窗口迭代器

- **处理器流水线**：`Processor` 抽象基类 + `Serializable` 序列化。关键处理器：
  - **去噪/清洗**：`DropnaProcessor`/`DropnaLabel`/`DropCol`/`FilterCol`/`ProcessInf`/`Fillna`
  - **标准化**：`MinMaxNorm`/`ZScoreNorm`/`RobustZScoreNorm`/`CSZScoreNorm`/`CSRankNorm`/`CSZFillna`（截面标准化，按日期分组）
  - **截面操作**：`CSZScoreNorm`/`CSRankNorm`/`CSZFillna`（按 `datetime` 分组 apply）
  - **格式转换**：`HashStockFormat` (DataFrame → HashingStockStorage)
  - **时间/工具过滤**：`TimeRangeFlt`/`FilterCol`/`DropCol`
  - **配置参数**：`fit_start_time`/`fit_end_time` 严格防 leakage（训练期不得包含测试期信息）

- **PIT (Point-in-Time) 算子**：`P`/`PRef` 算子
  - `P`：将 `period_time` 维度特征在 `observe_time t` 坍缩为单点，**只用 `period_time ≤ t` 的数据，未来数据物理不可达**
  - `PRef`：指定回看期的 P 算子
  - 表达式引擎层面的硬闸，非事后检查

## 我们现在怎么做的
`src/clawock/market_data/factors.py:314-400` — `FactorFetch._compute_cross_sectional()` 做简易截面排名：
```python
def _compute_cross_sectional(self, df: pd.DataFrame, fields: list[str]) -> pd.DataFrame:
    """按日期分组做 cross-sectional rank/zscore"""
    for field in fields:
        df[f"{field}_cs_rank"] = df.groupby("date")[field].rank(pct=True)
        df[f"{field}_cs_zscore"] = df.groupby("date")[field].transform(
            lambda x: (x - x.mean()) / x.std() if x.std() > 0 else 0
        )
    return df
```
**仅两行 pandas groupby**，无标准化器抽象、无 fit/transform 分离、无 leakage 防护。

`src/clawock/market_data/factors.py:450-550` — `FactorFetch._compute_rolling()` 简易滚动：
```python
def _compute_rolling(self, df: pd.DataFrame, fields: list[str], windows: list[int]) -> pd.DataFrame:
    for field in fields:
        for w in windows:
            df[f"{field}_mean_{w}"] = df.groupby("ticker")[field].transform(
                lambda x: x.rolling(w, min_periods=1).mean()
            )
```
**硬编码 rolling mean**，无统一滚动算子接口、无 Slope/Rsquare/Resi 等高级算子、无 pair-wise rolling (Corr/Cov)。

`src/clawock/market_data/factors.py:1-100` — 无表达式引擎，特征定义硬编码在 Python 代码里，无法声明式组合 `$close/Ref($close,1)-1` 这种表达式。

`src/clawock/market_data/factors.py` 全文 — 无 `Processor` 抽象、无 fit/transform 分离、无 `fit_start_time`/`fit_end_time` 防 leakage、无三层数据（DK_R/DK_I/DK_L）分离学习/推理预处理。

## 差在哪
**机制缺失**：
1. **无表达式引擎** — 无法声明式定义 `$close/Ref($close,1)-1`、`Mean($close,20)`、`Slope($close,10)/$close` 等复合特征；特征定义硬编码在 Python 里
2. **无统一算子注册机制** — 无法扩展自定义算子（如 qlib 的 `OpsWrapper.register()`），特征计算硬编码
3. **无 Processor 抽象** — 无 `fit`/`transform` 分离，无 `fit_start_time`/`fit_end_time` 防 leakage，无 `is_for_infer`/`readonly` 标记
3. **无截面处理器抽象** — `CSZScoreNorm`/`CSRankNorm`/`CSZFillna` 这种按日期分组的标准化/排名/填充，我们只有简易 `groupby().rank()`
4. **无三层数据分离** — 无 DK_R(原始)/DK_I(推理)/DK_L(学习) 三层数据隔离，学习/推理预处理混用
5. **无 PIT 算子** — 无 `P`/`PRef` 点在时间坍缩，表达式引擎层面无未来数据物理屏蔽
5. **无数据处理器统一接口** — 无 `DataHandlerLP` 这种支持 `proc_func` 预处理钩子、三层数据、滑动窗口迭代器的统一 fetch

## 改法
新建/改动文件（≤350 LOC，纯离线，无新增外部服务、无新增 cron）：

1. **新建 `src/clawock/features/expression.py`**（~120 LOC）
   - `Expression` 抽象基类：`evaluate(df: pd.DataFrame) -> pd.Series`、`depends_on() -> list[str]`、`window_size() -> int`
   - `LiteralExpr`/`ColumnExpr`/`UnaryExpr`/`BinaryExpr`/`RollingExpr`/`RefExpr`/`IfExpr`/`ResampleExpr` 具体节点
   - `ExpressionParser`：解析字符串表达式 `$close/Ref($close,1)-1` → AST
   - `ExpressionEngine`：`compile(expr_str) -> Expression`、`evaluate(expr, df) -> pd.Series`，内置算子注册表（`RollingExpr` 支持 `Sum`/`Mean`/`Std`/`Slope`/`Rsquare`/`Resi`/`Corr`/`Cov`/`Rank`/`Quantile`/`EMA`/`WMA`/`Delta`/`Min`/`Max`/`Ref`）

2. **新建 `src/clawock/features/processor.py`**（~100 LOC）
   - `Processor` 基类：`fit(df) -> None`、`transform(df) -> pd.DataFrame`、`is_for_infer() -> bool`、`readonly() -> bool`、`config(**kwargs)`
   - 核心处理器：`DropnaProcessor`/`DropnaLabel`/`Fillna`/`ProcessInf`/`DropCol`/`FilterCol`
   - 标准化器：`MinMaxNorm`/`ZScoreNorm`/`RobustZScoreNorm`（带 `fit_start_time`/`fit_end_time` 防 leakage）、**`CSZScoreNorm`/`CSRankNorm`/`CSZFillna` 截面标准化/排名/填充（按 `datetime` 分组）**
   - `ProcessorPipeline`：`fit(df)` 顺序调用每个 processor 的 `fit`，`transform(df)` 顺序调用 `transform`

3. **新建 `src/clawock/features/handler.py`**（~80 LOC）
   - `DataHandler` 基类：`fetch(selector, level, col_set, data_key, proc_func) -> pd.DataFrame`
   - `DataHandlerLP`：三层数据 `raw`/`infer`/`learn`，`fetch(data_key="raw|infer|learn")`，`proc_func` 预处理钩子
   - `get_range_iterator(periods)` 滑动窗口迭代器

4. **改 `src/clawock/market_data/factors.py`**（~60 LOC 增量）
   - `_compute_cross_sectional()` 改用 `CSZScoreNorm`/`CSRankNorm` 处理器
   - `_compute_rolling()` 改用 `ExpressionEngine` 编译 `RollingExpr`
   - 新增 `build_feature_pipeline(config)` 从配置构建 `ProcessorPipeline`

**规模量级**：~360 LOC 新增/改动，纯 Python，依赖 `pandas`/`numpy`（已在 `pyproject.toml`）。

## 验收
今天会失败、改完会通过的 pytest 断言（贴在测试文件 `tests/test_expression_engine.py`）：

```python
def test_expression_parser_compiles():
    from clawock.features.expression import ExpressionParser, ExpressionEngine
    parser = ExpressionParser()
    expr = parser.parse("$close / Ref($close, 1) - 1")
    assert expr is not None
    engine = ExpressionEngine()
    # 构造测试数据
    import pandas as pd
    import numpy as np
    dates = pd.date_range("2025-01-01", periods=10)
    df = pd.DataFrame({"close": np.arange(100, 110)}, index=dates)
    df.index.name = "date"
    result = engine.evaluate(expr, df)
    assert len(result) == 10
    # 第一个值应为 NaN（Ref 需要前值）
    assert pd.isna(result.iloc[0])
    # 第二个值：(101/100 - 1) = 0.01
    assert abs(result.iloc[1] - 0.01) < 1e-6

def test_rolling_expr_slope():
    from clawock.features.expression import ExpressionParser, ExpressionEngine
    parser = ExpressionParser()
    expr = parser.parse("Slope($close, 5) / $close")
    engine = ExpressionEngine()
    dates = pd.date_range("2025-01-01", periods=20)
    df = pd.DataFrame({"close": np.linspace(100, 120, 20)}, index=dates)
    df.index.name = "date"
    result = engine.evaluate(expr, df)
    assert len(result) == 20
    # 线性增长序列，斜率为正常数
    assert result.iloc[-1] > 0

def test_cs_zscore_norm_prevents_leakage():
    from clawock.features.processor import CSZScoreNorm
    import pandas as pd
    import numpy as np
    dates = pd.date_range("2025-01-01", periods=5)
    tickers = ["A", "B", "C"]
    idx = pd.MultiIndex.from_product([dates, tickers], names=["date", "ticker"])
    df = pd.DataFrame({"feature": np.random.randn(15)}, index=idx)
    processor = CSZScoreNorm(fields_group="feature")
    # 无 fit 调用直接 transform（截面标准化无需 fit）
    result = processor.transform(df.copy())
    # 每日截面均值约为 0，标准差约为 1
    for date in dates:
        daily = result.xs(date, level="date")["feature"]
        assert abs(daily.mean()) < 1e-6
        assert abs(daily.std() - 1.0) < 1e-6

def test_minmax_norm_prevents_leakage():
    from clawock.features.processor import MinMaxNorm
    import pandas as pd
    import numpy as np
    dates = pd.date_range("2025-01-01", periods=20)
    df = pd.DataFrame({"feature": np.linspace(0, 100, 20)}, index=dates)
    # 只用前 10 天 fit
    processor = MinMaxNorm(fit_start_time="2025-01-01", fit_end_time="2025-01-10", fields_group="feature")
    processor.fit(df)
    result = processor.transform(df.copy())
    # fit 期内归一化到 [0,1]
    assert result.loc["2025-01-01", "feature"] == 0.0
    assert result.loc["2025-01-10", "feature"] == 1.0
    # fit 期外也按 fit 期参数归一化（不 leakage）
    assert result.loc["2025-01-15", "feature"] == 1.5  # (150-0)/(100-0) = 1.5

def test_processor_pipeline_fit_transform():
    from clawock.features.processor import ProcessorPipeline, Fillna, ZScoreNorm, DropnaProcessor
    import pandas as pd
    import numpy as np
    df = pd.DataFrame({"a": [1, 2, np.nan, 4], "b": [10, 20, 30, 40]}, index=pd.date_range("2025-01-01", periods=4))
    pipeline = ProcessorPipeline([
        Fillna(fields_group=None, fill_value=0),
        ZScoreNorm(fit_start_time="2025-01-01", fit_end_time="2025-01-02", fields_group="a"),
        DropnaProcessor(fields_group="a"),
    ])
    pipeline.fit(df)
    result = pipeline.transform(df.copy())
    # 应该无 NaN，a 列已标准化
    assert not result.isna().any().any()

def test_data_handler_lp_three_layers():
    from clawock.features.handler import DataHandlerLP
    import pandas as pd
    import numpy as np
    dates = pd.date_range("2025-01-01", periods=10)
    tickers = ["A", "B"]
    idx = pd.MultiIndex.from_product([dates, tickers], names=["date", "ticker"])
    df = pd.DataFrame({"feature": np.random.randn(20), "label": np.random.randn(20)}, index=idx)
    handler = DataHandlerLP(instruments=tickers, start_time="2025-01-01", end_time="2025-01-10")
    # 模拟 data_loader.load 返回 df
    handler._data = df
    # 三层数据键
    raw = handler.fetch(data_key="raw")
    infer = handler.fetch(data_key="infer")
    learn = handler.fetch(data_key="learn")
    assert len(raw) == 20
    # proc_func 钩子
    def add_one(df):
        df = df.copy()
        df["feature"] += 1
        return df
    infer_proc = handler.fetch(data_key="infer", proc_func=add_one)
    assert (infer_proc["feature"] - infer["feature"] == 1).all()
```

**Comment by KCNyu** [comment]
## 关闭理由:产品边界之外(不是"以后再说")

本条由 `competitor-port` 自治循环生成,格式合规、机制描述属实,但它要移植的能力**落在 clawock 的产品定义之外**。

clawock 的终态是 **agent-native 决策工作流插件包 + 可验证 harness**:被外部 agent runtime
(OpenClaw / Claude Code / Codex …) 调用,负责「把证据和对立观点逼进一次决策 / 确定性对账 /
追踪 thesis-decision-execution-outcome / 把已量到的结果变成有界、可评审、可回滚的工作流改进」。
它明确**不做**模型客户端、不做 agent runtime、不做交易执行/做市/清算/合规申报基础设施。

本条要求的能力属于以下之一,均非上述边界内:

- **执行/微结构/做市/SOR/TCA/结算成本** — clawock 不下单、没有 orderbook、没有 venue 连接,
  移植过来的模型没有任何真实输入可喂,只能拿合成数据自证。
- **加密/DeFi/期权/期货/FX 等新资产类** — 需要数据适配器 + 每类的结算守恒不变量 + 重新预注册
  研究契约,是产品边界决策而不是代码任务(同 #1114 留开的那半)。
- **合规申报 / RegTech / KYC-AML** — 面向受监管机构的申报义务,clawock 没有这个身份。
- **重型 ML 基础设施(基础模型/GNN/RL/MLOps/特征商店/流处理/分布式回测)** — 引入的依赖体量与
  运行形态和「装进任何 agent 的插件包」直接冲突;本仓库的研究闸一贯是**样本量与预注册**,
  不是模型容量,新增一个可训练模型只会多一个过拟合面。

关掉不等于否认机制价值。同族里**在边界内**的那几条已单独留着并在实现:回测严谨性
(purged CSCV + PBO,`evaluation/cscv.py` 已上线)、可复现性指纹、因子/信号截面评估
(`clawock signal-panel`)。如果将来产品边界真的扩到某个资产类或执行层,重开一条**带当时的数据
适配器与守恒不变量方案**的 issue,比复活这条更省事。

参考:`clawock-product-goal`(2026-08-08 与 kcn 核对的终态定义)。

**Comment by KCNyu** [comment]
## 重开:关早了 —— 执行层不做,但**能辅助决策的因子要做**

kcn 更正口径:「可以不做执行层,但是有因子可以辅助决策的要做」。

上一条批量关闭把**执行/清算/合规/重型 ML 基础设施**和**这条 issue 里可以变成因子的那部分**
一起关了,这是我的取舍过度。执行层的判断不变(clawock 不下单、没有 orderbook、没有 venue 连接),
但本条里**只用现有日线/快照就能算出来、且能进决策的量**属于产品边界内的因子层,重开并做**限定实现**。

重开后本条的范围**收窄为**:

- ✅ 只用 `assets/data/*` 与 `memory/bars` 已有的数据算得出来的量;
- ✅ 产出进因子/信号层(`market_data/factors`、`evaluation/signal_panel`、研究面),
  按本仓库既有闸走(预注册、样本量、`usable_for_decisions` 只在证据够时才翻);
- ❌ 不做下单/路由/撮合/做市/清算/申报;
- ❌ 不引入新的重型依赖(本仓库硬依赖只有 `requests`,numpy 是可选);
- ❌ 不新增每日 cron、不碰 live 决策契约、不碰 200KB payload 闸。

原 issue 正文里超出这个范围的部分视为已关闭,不再是验收条件。

**Comment by KCNyu** [comment]
Closing: two of the three asks already exist, and the third is the one that would hurt.

**Cross-sectional processors — shipped.** `market_data/factors.py` is not "简易截面排名": its module docstring opens with "Point-in-time, sector-neutral cross-sectional factor research", and the pipeline is `winsorize_rows()` (`:355`) → per-sector ranking of every raw factor (`:482-501`, emitted as `sector_neutral_ranks`) → weighted composite with `factor_coverage_pct` attached. Winsorization and sector neutralisation are exactly the qlib processors this issue asks for.

**Point-in-time — shipped, in two layers.** `factors.py:187-212` reads PIT XBRL from SEC (revenue / gross profit / operating income matched on PIT semantics), and every registered signal history is written once per session and never rewritten, which is what makes `evaluation/signal_panel.py` able to score forward returns without reconstructing history.

**The expression engine is the ask I do not want to grant.** A factor DSL's value is that it makes new factors cheap; the published selection numbers say cheap factors are the problem, not the solution — `decision_map.json`'s `signal_panel` block reports selection PBO **0.81 at t1** over 70 splits, i.e. at the short horizon the winner is already more likely than not a product of choosing among the candidates we have. Nine registered factors, each with a forward-return score and an interval, is a deliberate ceiling on trials. A DSL raises that ceiling without adding evidence.

Same reasoning as #1166, which was the adjacent-labelled version of this.
